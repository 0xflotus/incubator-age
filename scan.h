#ifndef AG_SCAN_H
#define AG_SCAN_H

// AG_TOKEN_NULL indicates the end of a scan. The name came from YY_NULL.
//
// AG_TOKEN_DECIMAL can be a decimal integer literal that does not fit in "int"
// type.
enum ag_token_type
{
    AG_TOKEN_NULL,
    AG_TOKEN_INTEGER,
    AG_TOKEN_DECIMAL,
    AG_TOKEN_STRING,
    AG_TOKEN_IDENTIFIER,
    AG_TOKEN_PARAMETER,
    AG_TOKEN_LT_GT,
    AG_TOKEN_LT_EQ,
    AG_TOKEN_GT_EQ,
    AG_TOKEN_DOT_DOT,
    AG_TOKEN_PLUS_EQ,
    AG_TOKEN_EQ_TILDE,
    AG_TOKEN_CHAR
};

// Fields in value field are used with the following types.
//
//     * c: AG_TOKEN_CHAR
//     * i: AG_TOKEN_INTEGER
//     * s: all other types except the types for c and i, and AG_TOKEN_NULL
//
// "int" type is chosen for value.i to line it up with Value in PostgreSQL.
//
// value.s is read-only because it points at an internal buffer and it changes
// for every ag_scanner_next_token() call. So, users who want to keep or modify
// the value need to copy it first.
struct ag_token
{
    enum ag_token_type type;
    union
    {
        char c;
        int i;
        const char *s;
    } value;
    int location;
};

// an opaque data structure encapsulating the current state of the scanner
typedef void *ag_scanner_t;

ag_scanner_t ag_scanner_create(const char *s);
void ag_scanner_destroy(ag_scanner_t scanner);
struct ag_token ag_scanner_next_token(ag_scanner_t scanner);

#endif
