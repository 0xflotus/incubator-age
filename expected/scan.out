LOAD 'agensgraph';
SET search_path TO ag_catalog;
--
-- multi-line comment
--
SELECT * FROM cypher($$
/*
 * multi-line comment
 */
RETURN 0
/**/
$$) AS t(a agtype);
 a 
---
 0
(1 row)

SELECT * FROM cypher($$
/* unterminated /* comment
RETURN 0
$$) AS t(a int);
ERROR:  unterminated /* comment at or near "/* unterminated /* comment
RETURN 0
"
LINE 2: /* unterminated /* comment
        ^
-- recover syntax highlighting */
--
-- single-line comment
--
SELECT * FROM cypher($$
// single-line
// comment
RETURN 0
$$) AS t(a agtype);
 a 
---
 0
(1 row)

--
-- decimal integer
--
SELECT * FROM cypher($$
RETURN 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype, i agtype, j agtype);
 a | b | c | d | e | f | g | h | i | j  
---+---+---+---+---+---+---+---+---+----
 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
(1 row)

SELECT * FROM cypher($$
RETURN 11, 22, 33, 44, 55, 66, 77, 88, 99
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype, i agtype);
 a  | b  | c  | d  | e  | f  | g  | h  | i  
----+----+----+----+----+----+----+----+----
 11 | 22 | 33 | 44 | 55 | 66 | 77 | 88 | 99
(1 row)

-- 2^31 - 1, 2^31
SELECT * FROM cypher($$
RETURN 2147483647, 2147483648
$$) AS t(a agtype, b agtype);
     a      |     b      
------------+------------
 2147483647 | 2147483648
(1 row)

--
-- octal integer
--
SELECT * FROM cypher($$
RETURN 00, 01, 02, 03, 04, 05, 06, 07, 010
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype, i agtype);
 a | b | c | d | e | f | g | h | i 
---+---+---+---+---+---+---+---+---
 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
(1 row)

SELECT * FROM cypher($$
RETURN 000, 011, 022, 033, 044, 055, 066, 077
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype);
 a | b | c  | d  | e  | f  | g  | h  
---+---+----+----+----+----+----+----
 0 | 9 | 18 | 27 | 36 | 45 | 54 | 63
(1 row)

-- 2^31 - 1, 2^31
SELECT * FROM cypher($$
RETURN 000000000000, 017777777777, 0020000000000
$$) AS t(a agtype, b agtype, c agtype);
 a |     b      |     c      
---+------------+------------
 0 | 2147483647 | 2147483648
(1 row)

-- 2^60 - 1, 2^64 - 1
SELECT * FROM cypher($$
RETURN 077777777777777777777, 01777777777777777777777
$$) AS t(a agtype, b agtype);
          a          |          b           
---------------------+----------------------
 1152921504606846975 | 1.84467440737096e+19
(1 row)

-- an invalid character after reading valid digits
SELECT * FROM cypher($$
RETURN 012345678
$$) AS t(a int);
ERROR:  invalid octal integer literal at or near "012345678"
LINE 2: RETURN 012345678
               ^
-- an invalid character after the leading "0"
SELECT * FROM cypher($$
RETURN 09
$$) AS t(a int);
ERROR:  invalid octal integer literal at or near "09"
LINE 2: RETURN 09
               ^
--
-- hexadecimal integer
--
SELECT * FROM cypher($$
RETURN 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype, i agtype, j agtype);
 a | b | c | d | e | f | g | h | i | j 
---+---+---+---+---+---+---+---+---+---
 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
(1 row)

SELECT * FROM cypher($$
RETURN 0xA, 0xB, 0xC, 0xD, 0xE, 0xF
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype, f agtype);
 a  | b  | c  | d  | e  | f  
----+----+----+----+----+----
 10 | 11 | 12 | 13 | 14 | 15
(1 row)

SELECT * FROM cypher($$
RETURN 0X00, 0X11, 0X22, 0X33, 0X44, 0X55, 0X66, 0X77, 0X88, 0X99
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype,
         f agtype, g agtype, h agtype, i agtype, j agtype);
 a | b  | c  | d  | e  | f  |  g  |  h  |  i  |  j  
---+----+----+----+----+----+-----+-----+-----+-----
 0 | 17 | 34 | 51 | 68 | 85 | 102 | 119 | 136 | 153
(1 row)

SELECT * FROM cypher($$
RETURN 0XAa, 0XBb, 0XCc, 0XDd, 0XEe, 0xFf
$$) AS t(a agtype, b agtype, c agtype, d agtype, e agtype, f agtype);
  a  |  b  |  c  |  d  |  e  |  f  
-----+-----+-----+-----+-----+-----
 170 | 187 | 204 | 221 | 238 | 255
(1 row)

-- 2^31 - 1, 2^31
SELECT * FROM cypher($$
RETURN 0x00000000, 0x7FFFFFFF, 0x080000000
$$) AS t(a agtype, b agtype, c agtype);
 a |     b      |     c      
---+------------+------------
 0 | 2147483647 | 2147483648
(1 row)

-- 10^18, 2^64 - 1
SELECT * FROM cypher($$
RETURN 0xde0b6b3a7640000, 0xffffffffffffffff
$$) AS t(a agtype, b agtype);
          a          |          b           
---------------------+----------------------
 1000000000000000000 | 1.84467440737096e+19
(1 row)

-- an invalid character after reading valid digits
SELECT * FROM cypher($$
RETURN 0xF~
$$) AS t(a int);
ERROR:  unexpected character at or near "~"
LINE 2: RETURN 0xF~
                  ^
-- an invalid character after the leading "0x"
SELECT * FROM cypher($$
RETURN 0x~
$$) AS t(a int);
ERROR:  invalid hexadecimal integer literal at or near "0x"
LINE 2: RETURN 0x~
               ^
-- "0x" followed by nothing
SELECT * FROM cypher($$
RETURN 0x
$$) AS t(a int);
ERROR:  invalid hexadecimal integer literal at or near "0x"
LINE 2: RETURN 0x
               ^
--
-- decimal
--
SELECT * FROM cypher($$
RETURN 03., 3.141592, .141592
$$) AS t(a agtype, b agtype, c agtype);
  a  |    b     |    c     
-----+----------+----------
 3.0 | 3.141592 | 0.141592
(1 row)

-- "0" and ".."
SELECT * FROM cypher($$
RETURN 0..
$$) AS t(a text, b text, c text);
ERROR:  syntax error at or near ".."
LINE 2: RETURN 0..
                ^
--
-- scientific notation
--
SELECT * FROM cypher($$
RETURN 3141592e-6, 3.141592E0, .3141592e+1
$$) AS t(a agtype, b agtype, c agtype);
    a     |    b     |    c     
----------+----------+----------
 3.141592 | 3.141592 | 3.141592
(1 row)

-- invalid exponent parts
SELECT * FROM cypher($$
RETURN 3141592e-
$$) AS t(a text);
ERROR:  invalid scientific notation literal at or near "3141592e-"
LINE 2: RETURN 3141592e-
               ^
SELECT * FROM cypher($$
RETURN 3.141592E
$$) AS t(a text);
ERROR:  invalid scientific notation literal at or near "3.141592E"
LINE 2: RETURN 3.141592E
               ^
SELECT * FROM cypher($$
RETURN .3141592e+
$$) AS t(a text);
ERROR:  invalid scientific notation literal at or near ".3141592e+"
LINE 2: RETURN .3141592e+
               ^
--
-- quoted string
--
-- a long string
SELECT * FROM cypher($$
RETURN " !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
$$) AS t(a agtype);
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      a                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 " !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ !#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
(1 row)

-- escape sequences
SELECT * FROM cypher($$
RETURN " \" \" ' \' ", ' \' \' " \" ', " / \/ \\ \b \f \n \r \t "
$$) AS t(a agtype, b agtype, c agtype);
       a       |       b       |             c             
---------------+---------------+---------------------------
 " \" \" ' ' " | " ' ' \" \" " | " / / \\ \b \f \n \r \t "
(1 row)

-- invalid escape sequence
SELECT * FROM cypher($$
RETURN "\a"
$$) AS t(a text);
ERROR:  invalid escape sequence at or near "\a"
LINE 2: RETURN "\a"
                ^
DETAIL:  Valid escape sequences are \", \', \/, \\, \b, \f, \n, \r, \t, \uXXXX, and \UXXXXXXXX.
-- Unicode escape sequences
SELECT * FROM cypher($$
RETURN "\u03A9 (GREEK CAPITAL LETTER OMEGA, U+03A9, Ω)",
       "\U0001d6e9 (MATHEMATICAL ITALIC CAPITAL THETA, U+1D6E9, 𝛩)",
       "\ud835\U0000DEF0 (MATHEMATICAL ITALIC CAPITAL OMICRON, U+1D6F0, 𝛰)",
       "\u002E (FULL STOP, U+002E, .)"
$$) AS t(a agtype, b agtype, c agtype, d agtype);
                      a                      |                          b                          |                           c                           |             d              
---------------------------------------------+-----------------------------------------------------+-------------------------------------------------------+----------------------------
 "Ω (GREEK CAPITAL LETTER OMEGA, U+03A9, Ω)" | "𝛩 (MATHEMATICAL ITALIC CAPITAL THETA, U+1D6E9, 𝛩)" | "𝛰 (MATHEMATICAL ITALIC CAPITAL OMICRON, U+1D6F0, 𝛰)" | ". (FULL STOP, U+002E, .)"
(1 row)

-- invalid Unicode surrogate pair (need a low surrogate)
SELECT * FROM cypher($$
RETURN "\uD835"
$$) AS t(a text);
ERROR:  invalid Unicode surrogate pair at or near """
LINE 2: RETURN "\uD835"
                      ^
DETAIL:  A low surrogate must follow a high surrogate.
-- invalid Unicode surrogate pair (not a low surrogate)
SELECT * FROM cypher($$
RETURN "\uD835\u002E"
$$) AS t(a text);
ERROR:  invalid Unicode surrogate pair at or near "\u002E"
LINE 2: RETURN "\uD835\u002E"
                      ^
DETAIL:  A low surrogate must follow a high surrogate.
-- invalid Unicode surrogate pair (need a high surrogate)
SELECT * FROM cypher($$
RETURN "\uDEF0"
$$) AS t(a text);
ERROR:  invalid Unicode surrogate pair at or near "\uDEF0"
LINE 2: RETURN "\uDEF0"
                ^
DETAIL:  A low surrogate must follow a high surrogate.
-- invalid Unicode escape value (must be less than or equal to 10FFFF)
SELECT * FROM cypher($$
RETURN "\U00110000"
$$) AS t(a text);
ERROR:  invalid Unicode escape value at or near "\U00110000"
LINE 2: RETURN "\U00110000"
                ^
DETAIL:  Unicode escape values cannot be greater than 10FFFF, which is the maximum value of a code point.
-- unsupported Unicode escape value ('\0' is not allowed)
SELECT * FROM cypher($$
RETURN "\u0000"
$$) AS t(a text);
ERROR:  unsupported Unicode escape value at or near "\u0000"
LINE 2: RETURN "\u0000"
                ^
DETAIL:  Unicode code point value 0000 is not allowed in quoted strings.
-- unsupported Unicode escape value (the server encoding is not UTF8)
CREATE DATABASE contrib_regression_agensgraph_euc_kr
  TEMPLATE template0
  ENCODING EUC_KR
  LC_COLLATE 'C' LC_CTYPE 'C';
\c contrib_regression_agensgraph_euc_kr
CREATE EXTENSION agensgraph;
LOAD 'agensgraph';
SET search_path TO ag_catalog;
SELECT * FROM cypher($$
RETURN "\U0001D706"
$$) AS t(a text);
ERROR:  unsupported Unicode escape value at or near "\U0001D706"
LINE 2: RETURN "\U0001D706"
                ^
DETAIL:  Unicode escape values cannot be used for code point values above 007F when the server encoding is not UTF8.
\c contrib_regression
DROP DATABASE contrib_regression_agensgraph_euc_kr;
LOAD 'agensgraph';
SET search_path TO ag_catalog;
-- invalid Unicode escape sequence (must be \uXXXX or \UXXXXXXXX)
SELECT * FROM cypher($$
RETURN "\UD835"
$$) AS t(a text);
ERROR:  invalid Unicode escape sequence at or near "\UD835"
LINE 2: RETURN "\UD835"
                ^
HINT:  Unicode escape sequences must be \uXXXX or \UXXXXXXXX.
SELECT * FROM cypher($$
RETURN "\uD835\uDEF"
$$) AS t(a text);
ERROR:  invalid Unicode escape sequence at or near "\uDEF"
LINE 2: RETURN "\uD835\uDEF"
                      ^
HINT:  Unicode escape sequences must be \uXXXX or \UXXXXXXXX.
-- unterminated quoted strings
SELECT * FROM cypher($$RETURN "unterminated quoted string$$) AS t(a text);
ERROR:  unterminated quoted string at or near ""unterminated quoted string"
LINE 1: SELECT * FROM cypher($$RETURN "unterminated quoted string$$)...
                                      ^
-- recover syntax highlighting "
SELECT * FROM cypher($$RETURN 'unterminated quoted string$$) AS t(a text);
ERROR:  unterminated quoted string at or near "'unterminated quoted string"
LINE 1: SELECT * FROM cypher($$RETURN 'unterminated quoted string$$)...
                                      ^
-- recover syntax highlighting '
SELECT * FROM cypher($$RETURN "escape \$$) AS t(a text);
ERROR:  unterminated quoted string at or near ""escape \"
LINE 1: SELECT * FROM cypher($$RETURN "escape \$$) AS t(a text);
                                      ^
-- recover syntax highlighting "
SELECT * FROM cypher($$RETURN "high surrogate \uD835$$) AS t(a text);
ERROR:  unterminated quoted string at or near ""high surrogate \uD835"
LINE 1: SELECT * FROM cypher($$RETURN "high surrogate \uD835$$) AS t...
                                      ^
-- recover syntax highlighting "
--
-- identifier
--
-- check that they are accepted as identifier (all tests throw an error)
SELECT * FROM cypher($$
RETURN _$09A_z
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
SELECT * FROM cypher($$
RETURN A
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
SELECT * FROM cypher($$
RETURN z
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
SELECT * FROM cypher($$
RETURN `$`
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
SELECT * FROM cypher($$
RETURN `0`
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
SELECT * FROM cypher($$
RETURN ````
$$) AS t(id text);
ERROR:  unrecognized node type: 341
LINE 1: SELECT * FROM cypher($$
                              ^
-- zero-length quoted identifier
SELECT * FROM cypher($$
RETURN ``
$$) AS t(a text);
ERROR:  zero-length quoted identifier at or near "``"
LINE 2: RETURN ``
               ^
SELECT * FROM cypher($$
RETURN `unterminated quoted identifier
$$) AS t(a text);
ERROR:  unterminated quoted identifier at or near "`unterminated quoted identifier
"
LINE 2: RETURN `unterminated quoted identifier
               ^
-- recover syntax highlighting `
--
-- parameter
--
-- invalid parameter names
SELECT * FROM cypher($cypher$
RETURN $$
$cypher$) AS t(a text);
ERROR:  unexpected character at or near "$"
LINE 2: RETURN $$
               ^
SELECT * FROM cypher($$
RETURN $0
$$) AS t(a text);
ERROR:  unexpected character at or near "$"
LINE 2: RETURN $0
               ^
